readthedocs/core/unresolver.py

```diff
@@ -1,99 +1,284 @@
+        If the subproject exists, we try to resolve the rest of the path
+        with the subproject as the canonical project.
+        """
+        match = self.subproject_pattern.match(path)
+        if not match:
             return None
 
-        _, __, kwargs = url_resolve(
-            path,
-            urlconf='readthedocs.proxito.urls',
+        project_slug = match.group("project")
+        file = self._normalize_filename(match.group("file"))
+        project_relationship = (
+            parent_project.subprojects.filter(alias=project_slug)
+            .prefetch_related("child")
+            .first()
         )
+        if project_relationship:
+            return self._unresolve_path(
+                parent_project=project_relationship.child,
+                path=file,
+                check_subprojects=False,
+            )
+        return None
 
-        mixin = ServeDocsMixin()
-        version_slug = mixin.get_version_from_host(request, kwargs.get('version_slug'))
+    def _match_single_version_project(self, parent_project, path):
+        """
+        Try to match a single version project.
 
-        final_project, lang_slug, version_slug, filename = _get_project_data_from_request(  # noqa
-            request,
-            project_slug=project_slug,
-            subproject_slug=kwargs.get('subproject_slug'),
-            lang_slug=kwargs.get('lang_slug'),
-            version_slug=version_slug,
-            filename=kwargs.get('filename', ''),
-        )
+        By default any path will match.
+        """
+        file = self._normalize_filename(path)
+        version = parent_project.versions.filter(
+            slug=parent_project.default_version
+        ).first()
+        if version:
+            return parent_project, version, file
+        return parent_project, None, None
 
-        # Handle our backend storage not supporting directory indexes,
-        # so we need to append index.html when appropriate.
-        if not filename or filename.endswith('/'):
-            # We need to add the index.html to find this actual file
-            filename += 'index.html'
-
-        log.debug(
-            'Unresolver parsed.',
-            project_slug=final_project.slug,
-            lang_slug=lang_slug,
-            version_slug=version_slug,
-            filename=filename,
+    def _unresolve_path(self, parent_project, path, check_subprojects=True):
+        """
+        Unresolve `path` with `parent_project` as base.
+
+        If the returned project is `None`, then we weren't able to
+        unresolve the path into a project.
+
+        If the returned version is `None`, then we weren't able to
+        unresolve the path into a valid version of the project.
+
+        :param parent_project: The project that owns the path.
+        :param path: The path to unresolve.
+        :param check_subprojects: If we should check for subprojects,
+         this is used to call this function recursively.
```

Comment:
This could expand about how it's used.

It seems we are calling this recursively _only_ for the subproject case. I guess this is because we want to know if the subproject is multiversion or single version, but it can't have subprojects since we don't support "subprojects on subprojects"